src/main/java/com/icthh/xm/ms/entity/service/impl/XmEntityServiceImpl.java

```diff
@@ -189,6 +198,83 @@ public XmEntity saveXmEntity(XmEntity xmEntity) {
         return xmEntityRepository.save(xmEntity);
     }
 
+    private void assertSourceEntityMaxLinkValue(XmEntity xmEntity) {
+        if (xmEntity.getSources().isEmpty()) {
+            return;
+        }
+        Set<Link> sources = xmEntity.getSources();
+        Set<Link> newLinks = sources.stream().filter(Link::isNew).collect(toSet());
+        List<Long> sourceIds = newLinks.stream().map(Link::getSource)
+            .map(XmEntity::getId).collect(toList());
+        List<String> sourceStateProjection = xmEntityRepository.findAllStateProjectionByIdIn(sourceIds)
+            .stream().map(XmEntityStateProjection::getTypeKey).collect(toList());
+        Map<Long, String> sourceIdsAndTypeKeys = IntStream.range(0, sourceIds.size()).boxed()
+            .collect(toMap(sourceIds::get,sourceStateProjection::get));
+
+        newLinks.forEach(it -> linkCount(it.getTypeKey(),
+                                            it.getSource(),
+                                            newLinks,
+                                            sourceIdsAndTypeKeys));
+    }
+
+    private void linkCount(String newLinkTypeKey,
+                           XmEntity source,
+                           Set<Link> newLinks,
+                           Map<Long, String> sourceIdsAndTypeKeys) {
+
+        Predicate<Link> checkTypeKeyPredicate = it -> it.getTypeKey().equals(newLinkTypeKey);
+        Long newLinkCount = newLinks.stream().filter(it -> it.getSource().getId().equals(source.getId()))
+            .filter(checkTypeKeyPredicate).count();
+        Long countOfSourceTarget = (long) linkRepository.countBySourceIdAndTypeKey(source.getId(),newLinkTypeKey);
```

Comment:
request in db for each new link?
even if it's link with same types and from same entities?